processor/src/chiplets/aux_trace/mod.rs

```diff
@@ -245,3 +304,995 @@ impl AuxColumnBuilder<ChipletsVTableUpdate, ChipletsVTableRow, u32> for Chiplets
+    let transition_label = if first_cycle_row { op_label + 16 } else { op_label + 32 };
+
+    let header = alphas[0]
+        + alphas[1].mul_base(Felt::from(transition_label))
+        + alphas[2].mul_base(addr_nxt - ONE)
+        + alphas[3].mul_base(ZERO);
+
+    let state = &main_trace.chiplet_hasher_state(i - 2)[4..];
+    let state_nxt = &main_trace.chiplet_hasher_state(i - 1)[4..];
+
+    header + build_value(&alphas[8..16], state_nxt) - build_value(&alphas[8..16], state)
+}
+
+/// Builds requests made to the hasher chiplet at the end of a block.
+fn build_end_block_request<E: FieldElement<BaseField = Felt>>(
+    main_trace: &MainTrace,
+    alphas: &[E],
+    i: usize,
+) -> E {
+    let op_label = RETURN_HASH_LABEL;
+    let addr = main_trace.addr(i) + Felt::from(7_u64);
+
+    let first_cycle_row = addr_to_row_index(addr) % 8 == 0;
+    let transition_label = if first_cycle_row { op_label + 16 } else { op_label + 32 };
+
+    let header =
+        alphas[0] + alphas[1].mul_base(Felt::from(transition_label)) + alphas[2].mul_base(addr);
+
+    let state = main_trace.decoder_hasher_state(i);
+    let digest = &state[..4];
+
+    header + build_value(&alphas[8..12], digest)
+}
+
+/// Builds requests made to the bitwise chiplet. This can be either a request for the computation
+/// of a `XOR` or an `AND` operation.
+fn build_bitwise_request<E: FieldElement<BaseField = Felt>>(
+    main_trace: &MainTrace,
+    is_xor: Felt,
+    alphas: &[E],
+    i: usize,
+) -> E {
+    let op_label = get_op_label(ONE, ZERO, is_xor, ZERO);
+    let a = main_trace.s0(i);
+    let b = main_trace.s1(i);
+    let z = main_trace.s0(i + 1);
+
+    alphas[0]
+        + alphas[1].mul_base(op_label)
+        + alphas[2].mul_base(b)
+        + alphas[3].mul_base(a)
+        + alphas[4].mul_base(z)
+}
+
+/// Builds `MLOADW` and `MSTOREW` requests made to the memory chiplet.
+fn build_mloadw_mstorew_request<E: FieldElement<BaseField = Felt>>(
+    main_trace: &MainTrace,
+    is_mstorew: bool,
+    alphas: &[E],
+    i: usize,
+) -> E {
+    let ctx = main_trace.ctx(i);
+    let clk = main_trace.clk(i);
+
+    let s0_cur = main_trace.s0(i);
+    let s0_nxt = main_trace.s0(i + 1);
+    let s1_nxt = main_trace.s1(i + 1);
+    let s2_nxt = main_trace.s2(i + 1);
+    let s3_nxt = main_trace.s3(i + 1);
+
+    let op_label = if is_mstorew {
+        MEMORY_WRITE_LABEL
+    } else {
+        MEMORY_READ_LABEL
+    };
+
+    alphas[0]
+        + alphas[1].mul_base(Felt::from(op_label))
+        + alphas[2].mul_base(ctx)
+        + alphas[3].mul_base(s0_cur)
+        + alphas[4].mul_base(clk)
+        + alphas[5].mul_base(s3_nxt)
+        + alphas[6].mul_base(s2_nxt)
+        + alphas[7].mul_base(s1_nxt)
+        + alphas[8].mul_base(s0_nxt)
+}
+
+/// Builds `MLOAD` and `MSTORE` requests made to the memory chiplet.
+fn build_mload_mstore_request<E: FieldElement<BaseField = Felt>>(
+    main_trace: &MainTrace,
+    is_mstore: bool,
+    alphas: &[E],
+    i: usize,
+) -> E {
```

Comment:
Similar comment to the one above, but I'm also wondering if we can combine these two functions into a single `build_mem_request` function. This would require a single conditional statement in the function like this:
```Rust
let (v0, v1, v2, v3) = if op_label == MEMORY_READ_LABEL || op_label == MEMORY_WRITE_LABEL {

} else {

};
```

Reply:
I am not sure I follow the conditions for the if-else statement.
I have implemented a version compressing the two functions into one, let me know if this does the job.

Reply:
Ah - for some reason I thought we have 4 different labels for reading/writing memory. But since we have only 2, your implementation makes sense.